keps/1282-pods-ready-requeue-strategy/README.md

```diff
@@ -75,25 +107,69 @@ Consider including folks who also work outside the SIG or subproject.
 
 ### API Changes
 
-Add an additional field to the Kueue ConfigMap to allow administrators to specify what timestamp to consider during queue sorting (under the pre-existing waitForPodsReady block).
+#### KueueConfig
+
+Add an additional field to the KueueConfig to allow administrators to specify what timestamp to consider during queue sorting (under the pre-existing waitForPodsReady block).
 
 Possible settings:
 
 * `Eviction` (Back of queue)
 * `Creation` (Front of queue)
 
-```yaml
-kind: ConfigMap
-metadata:
-  name: kueue-manager-config
-  namespace: kueue-system
-data:
-  controller_manager_config.yaml: |
-    apiVersion: config.kueue.x-k8s.io/v1beta1
-    kind: Configuration
-    # ...
-    waitForPodsReady:
-      requeuingTimestamp: Creation | Eviction # <-- New field
+```go
+type WaitForPodsReady struct {
+	...
+	// requeuingStrategy defines the strategy for requeuing a Workload
+	// +optional
+	RequeuingStrategy *RequeuingStrategy `json:"requeuingStrategy,omitempty"`
+}
+
+type RequeuingStrategy struct {
+	// timestamp defines the timestamp used for requeuing a Workload
+	// that was evicted due to Pod readiness. Defaults to Eviction.
+	// +optional
+	Timestamp *RequeuingTimestamp `json:"timestamp,omitempty"`
+	
+	// backOffLimitCount defines the maximum number of requeuing retries.
+	// When the number is reached, the workload is deactivated.	
+	//
+	// Defaults to null. 
+	// +optional
+	BackOffLimitCount *int32 `json:"backOffLimitCount,omitempty"`
+    
+	// backOffLimitTimeout defines the time for a workload that 
+	// has once been admitted to reach the PodsReady=true condition. 
+	// When the time is reached, the workload is deactivated.
+	// 	
+	// Defaults to null.
+	// +optional
+	BackOffLimitTimeout *int32 `json:"backOffLimitTimeout,omitempty"`
+}
+
+type RequeuingTimestamp string
+
+const (
+	// creationTimestamp timestamp (from Workload .metadata.creationTimestamp).
+	CreationTimestamp RequeuingTimestamp = "Creation"
+    
+	// evictionTimestamp timestamp (from Workload .status.conditions).
+	EvictionTimestamp RequeuingTimestamp = "Eviction"
+)
+```
+
+#### Workload
+
+Add a new field, "requeuedCount", to the Workload to allow recording the number of times a workload is requeued.
+
+```go
+type WorkloadStatus struct {
+	...
+	// requeuedCount records the number of times a workload has been requeued.
+	// When a deactivated workload is reactivated, this count is reset to 0. 
+	//
+	// +optional
+	RequeuedCount *int32 `json:"requeuedCount,omitempty"`
```

Comment:
What are the invariants that make a workload in "backoff" state? Is it a combination of `PodsReadyTimeout` condition and `.status.requeuedCount > 0`?

Reply:
Yes, that's right. Those parameters aren't configurable.

Reply:
Ah, there is one more thing. We won't expose the parameters to initialize Backoff: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait@v0.29.1#Backoff